_posts/2024-01-29-quarkus-saas-backend-time-tracking-app-case-study.adoc

```diff
@@ -0,0 +1,140 @@
+Fortunately, Quarkus has excellent Flyway support, so all our database migrations were in one place and executed during our back-end booting process.
+This kept our database schema and data transparent and reproducible.
+
+.Using Flyway to execute database migrations
+image::db-migrations.png[alt="Database Migrations",align="center",width=640]
+
+For our deployments, we use Kubernetes.
+Before using Quarkus, we described our application requirements using helm packaging, but with Quarkus, we opted for another approach as Quarkus offers a great Kubernetes extension.
+Instead of writing any code, we described our Kubernetes resources using an application.yaml file, keeping our complete application configuration in one place.
+This extension generated Kubernetes resource files behind the scenes, which we then applied to our Kubernetes cluster.
+This works well for us.
+
+.Using the Kubernetes extension to generate Kubernetes resources
+image::k8s-config.png[alt="Kubernetes configuration",align="center",width=640]
+
+For packaging our back-end API, we used the Jib extension.
+To package our application in a container, all we had to do was use the application.yaml file and set all the required parameters such as image name tags repository, and so on.
+We didn't have to maintain the Docker file on our own, which was very convenient.
+
+Our time-tracking application needs to send emails to our users and admins on various occasions.
+To keep things simpler, we decided not to go for any third-party API-driven email-sending approach.
+Instead, we send emails ourselves, and for that purpose, we use Qute email templates, which make composing and sending emails to our users very simple.
+This extension provides support for coding coroutines, allowing for non-blocking sending and higher throughput.
+
+.Using Qute email templates to send emails
+image::qute-templates.png[alt="Qute Templates",align="center",width=640]
+
+== Development Journey
+
+The Quarkus development process has been excellent so far.
+Compared to other frameworks like Spring Boot, Quarkus has a faster startup time and a smaller memory footprint.
+It also provides profiles, which allows us to have slightly different configurations or behaviors between environments.
+We can easily substitute some hard-to-run third-party services with local mocks, leaving the application code unchanged.
+Quarkus is also great in terms of configuration and how easily we can overwrite values stored in the application.yaml file with external environment variables.
+Although the hot reload mode didn't work well with Kotlin, I believe all the bugs related to it will be solved in upcoming releases.
+During development, we had to restart our running service most of the time for code changes to take effect.
+
+Our backend API functionalities took approximately a month and a half to complete.
+Considering that only two developers worked on the backend, I think it was an okay result.
```

Comment:
okey? is that a bad okey or a good okay ? :)

Reply:
@maxandersen I updated the wording slightly. Yes, the result was good, indeed